File: src/api/portfolioService.ts
const API_URL = import.meta.env.VITE_API_URL || 'http://localhost:8000';

export interface PortfolioHistoryPoint {
    timestamp: string;
    value: number;
    unrealized_pnl?: number;
    realized_pnl?: number;
}

export interface PortfolioMetrics {
    portfolio_value: number;
    cash_balance: number;
    position_value: number;
    unrealized_pnl: number;
    realized_pnl: number;
}

export type PortfolioTimeframe = '1d' | '7d' | '30d' | 'all';

export interface FundingHistoryData {
    id: string;
    type: 'Deposit' | 'Withdraw';
    asset: string;
    amount: number;
    txHash: string;
    status: 'Completed' | 'Pending' | 'Failed';
    timestamp: string;
}

export const portfolioService = {
    async getPortfolioHistory(
        userAddress: string,
        timeframe: PortfolioTimeframe = '1d',
        limit: number = 500
    ): Promise<{ data: PortfolioHistoryPoint[] }> {
        const params = new URLSearchParams({
            timeframe,
            limit: limit.toString()
        });

        const response = await fetch(
            `${API_URL}/api/portfolio/${userAddress}/history?${params.toString()}`
        );

        if (!response.ok) {
            throw new Error('Failed to fetch portfolio history');
        }

        return response.json();
    },

    async getCurrentPortfolioValue(userAddress: string): Promise<PortfolioMetrics> {
        const response = await fetch(
            `${API_URL}/api/portfolio/${userAddress}/current`
        );

        if (!response.ok) {
            throw new Error('Failed to fetch current portfolio value');
        }

        return response.json();
    },

    async createSnapshot(userAddress: string): Promise<{ status: string; message: string }> {
        const response = await fetch(
            `${API_URL}/api/portfolio/${userAddress}/snapshot`,
            {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' }
            }
        );

        if (!response.ok) {
            throw new Error('Failed to create portfolio snapshot');
        }

        return response.json();
    },

    // Funding History
    async getFundingHistory(userAddress: string, type?: 'Deposit' | 'Withdraw'): Promise<{ data: FundingHistoryData[] }> {
        const params = new URLSearchParams();
        if (type) params.append('type', type);

        try {
            const response = await fetch(
                `${API_URL}/api/portfolio/${userAddress}/funding?${params.toString()}`
            );

            if (!response.ok) {
                console.warn("Failed to fetch funding history");
                return { data: [] };
            }

            return response.json();
        } catch (error) {
            console.error("Error fetching funding history:", error);
            // Return empty array on error to prevent UI crash
            return { data: [] };
        }
    },

    // Trade History
    async getTradeHistory(userAddress: string): Promise<{ data: any[] }> {
        try {
            const response = await fetch(
                `${API_URL}/api/portfolio/${userAddress}/trades`
            );

            if (!response.ok) {
                console.warn("Failed to fetch trade history");
                return { data: [] };
            }

            return response.json();
        } catch (error) {
            console.error("Error fetching trade history:", error);
            return { data: [] };
        }
    }
};

File: websocket/services/portfolio_service.py
"""
Portfolio Service - Calculate and track portfolio value over time
"""
from datetime import datetime, timedelta
import os
from sqlalchemy.orm import Session
from sqlalchemy import func
from typing import List, Dict, Optional
import logging

from database.models import Position, Order, PortfolioSnapshot, FundingHistory, LedgerAccount
from database.connection import get_db

logger = logging.getLogger(__name__)

from sqlalchemy import select

class PortfolioService:
    """Service for calculating and managing portfolio values"""
    
    def __init__(self, db: Session):
        self.db = db
    
    async def calculate_portfolio_value(self, user_address: str) -> Dict[str, float]:
        """
        Calculate current portfolio value for a user using Ledger State
        """
        logger.info(f"Calculating portfolio value for {user_address}")
        
        # 1. Get Balance from Ledger
        result = await self.db.execute(select(LedgerAccount).where(LedgerAccount.address == user_address.lower()))
        ledger = result.scalar_one_or_none()

        force_mode = os.getenv("FORCE_EXECUTION_MODE", "auto").lower().strip()
        if ledger:
            realized_pnl = float(ledger.realized_pnl or 0.0)
            if force_mode == "simulation":
                initial_simulation_balance = float(
                    os.getenv("INITIAL_SIMULATION_BALANCE", "1000")
                )
                cash_balance = initial_simulation_balance + realized_pnl
            else:
                cash_balance = float(ledger.balance or 0.0)
            locked_margin = float(ledger.locked_margin or 0.0)
        else:
            cash_balance = 0.0
            locked_margin = 0.0
            realized_pnl = 0.0
        
        # 2. Get all open positions
        result = await self.db.execute(
            select(Position).where(
                Position.user_address == user_address.lower(),
                Position.status == 'OPEN'
            )
        )
        positions = result.scalars().all()
        
        position_value = sum(p.margin_used for p in positions)
        unrealized_pnl = sum(p.unrealized_pnl for p in positions)
        
        portfolio_value = cash_balance + unrealized_pnl
        
        return {
            'portfolio_value': portfolio_value,
            'cash_balance': cash_balance - locked_margin,
            'locked_margin': locked_margin,
            'position_value': position_value,
            'unrealized_pnl': unrealized_pnl,
            'realized_pnl': realized_pnl
        }
    
    async def save_snapshot(self, user_address: str):
        """
        Save current portfolio snapshot to database
        """
        try:
            metrics = await self.calculate_portfolio_value(user_address)
            
            snapshot = PortfolioSnapshot(
                user_address=user_address.lower(),
                timestamp=datetime.utcnow(),
                portfolio_value=metrics['portfolio_value'],
                cash_balance=metrics['cash_balance'],
                position_value=metrics['position_value'],
                unrealized_pnl=metrics['unrealized_pnl'],
                realized_pnl=metrics['realized_pnl']
            )
            
            self.db.add(snapshot)
            self.db.commit()
            
            logger.info(f"Saved portfolio snapshot for {user_address}: ${metrics['portfolio_value']:.2f}")
            
        except Exception as e:
            logger.error(f"Failed to save portfolio snapshot for {user_address}: {e}")
            self.db.rollback()
    
    async def get_portfolio_history(
        self,
        user_address: str,
        timeframe: str = '1d',
        limit: int = 500
    ) -> List[Dict]:
        """
        Get portfolio value history for charting
        
        Args:
            user_address: User wallet address
            timeframe: '1d', '7d', '30d', 'all'
            limit: Max number of data points
        
        Returns:
            List of {timestamp, value} dicts
        """
        # Calculate time range
        cutoff_time = self._get_cutoff_time(timeframe)
        
        stmt = select(PortfolioSnapshot).where(
            func.lower(PortfolioSnapshot.user_address) == user_address.lower()
        )
        
        if cutoff_time:
            stmt = stmt.where(PortfolioSnapshot.timestamp >= cutoff_time)
        
        stmt = stmt.order_by(PortfolioSnapshot.timestamp).limit(limit)
        
        result = await self.db.execute(stmt)
        snapshots = result.scalars().all()
        
        return [
            {
                'timestamp': s.timestamp.isoformat(),
                'value': s.portfolio_value,
                'unrealized_pnl': s.unrealized_pnl,
                'realized_pnl': s.realized_pnl
            }
            for s in snapshots
        ]

    async def get_funding_history(self, user_address: str, type: Optional[str] = None) -> List[Dict]:
        """
        Get deposit/withdraw history
        """
        stmt = select(FundingHistory).where(
            func.lower(FundingHistory.user_address) == user_address.lower()
        )

        if type:
            stmt = stmt.where(FundingHistory.type == type)

        stmt = stmt.order_by(FundingHistory.timestamp.desc())
        
        result = await self.db.execute(stmt)
        history = result.scalars().all()

        return [
            {
                "id": str(h.id),
                "type": h.type,
                "asset": h.asset,
                "amount": h.amount,
                "txHash": h.tx_hash,
                "status": h.status,
                "timestamp": h.timestamp.isoformat()
            }
            for h in history
        ]

    async def create_funding_record(self, user_address: str, type: str, asset: str, amount: float, tx_hash: str, status: str = 'Completed'):
        """Record a new deposit or withdrawal"""
        try:
            record = FundingHistory(
                user_address=user_address.lower(),
                type=type,
                asset=asset,
                amount=amount,
                tx_hash=tx_hash,
                status=status
            )
            self.db.add(record)
            await self.db.commit()
            return record
        except Exception as e:
            logger.error(f"Failed to save funding record: {e}")
            await self.db.rollback()
            raise
    
    async def _calculate_realized_pnl(self, user_address: str) -> float:
        """
        Calculate realized PNL from closed positions
        """
        stmt = select(
            func.sum(
                func.case(
                    (Order.side == 'sell', Order.notional_usd),
                    else_=-Order.notional_usd
                )
            )
        ).where(
            func.lower(Order.user_address) == user_address.lower(),
            Order.status == 'filled'
        )
        
        result = await self.db.execute(stmt)
        return float(result.scalar() or 0)
    
    def _get_cutoff_time(self, timeframe: str) -> Optional[datetime]:
        """Get cutoff datetime for given timeframe"""
        timeframes = {
            '1d': timedelta(days=1),
            '7d': timedelta(days=7),
            '30d': timedelta(days=30),
            'all': None
        }
        
        delta = timeframes.get(timeframe)
        if delta is None:
            return None
        
        return datetime.utcnow() - delta
    
    async def snapshot_all_active_users(self):
        """
        Create snapshots for all users with open positions
        Called by periodic background task
        """
        # Get unique users with open positions
        stmt = select(Position.user_address).distinct()
        result = await self.db.execute(stmt)
        active_users = result.all()
        
        logger.info(f"Creating snapshots for {len(active_users)} active users")
        
        for (user_address,) in active_users:
            await self.save_snapshot(user_address)
        
        logger.info("Portfolio snapshots completed")
